mlflow/transformers.py

```diff
@@ -0,0 +1,793 @@
+    :return: A string representing the engine type of the model being used (one of 'tf', 'flax', or
+             'torch').
+    """
+    engine = _get_engine_type(model)
+    if engine == "torch":
+        return ["torch", "torchvision"]
+    else:
+        return [engine]
+
+
+def get_default_pip_requirements(model) -> List[str]:
+    """
+    :param model: The model instance to be saved in order to provide the required underlying
+                  deep learning execution framework dependency requirements. Note that this must
+                  be the actual model instance and not a Pipeline.
+    :return: A list of default pip requirements for MLflow Models that have been produced with the
+             ``transformers`` flavor. Calls to :py:func:`save_model()` and :py:func:`log_model()`
+             produce a pip environment that contain these requirements at a minimum.
+    """
+    from transformers import TFPreTrainedModel, FlaxPreTrainedModel, PreTrainedModel
+
+    if not isinstance(model, (TFPreTrainedModel, FlaxPreTrainedModel, PreTrainedModel)):
+        raise MlflowException(
+            "The supplied model type is unsupported. The model must be one of: PreTrainedModel, "
+            "TFPreTrainedModel, or FlaxPreTrainedModel"
+        )
+    base_reqs = ["transformers"]
+    base_reqs.extend(_model_engine_type(model))
+    return [_get_pinned_requirement(module) for module in base_reqs]
+
+
+def get_default_conda_env(
+    model,
+) -> Union[None, Dict[str, Union[str, List[Union[str, Dict[str, List[str]]]]]]]:
+    """
+    :return: The default Conda environment for MLflow Models produced with the ``transformers``
+             flavor, based on the model instance framework type of the model to be logged.
+    """
+    return _mlflow_conda_env(additional_pip_deps=get_default_pip_requirements(model))
+
+
+@format_docstring(LOG_MODEL_PARAM_DOCS.format(package_name=FLAVOR_NAME))
+def save_model(
+    transformers_model,
+    path: str,
+    processor=None,
+    task: Optional[str] = None,
+    model_card=None,
+    conda_env: Optional[
+        Union[str, Dict[str, Union[str, List[Union[str, Dict[str, List[str]]]]]]]
+    ] = None,
+    code_paths: Optional[List[str]] = None,
+    mlflow_model: Optional[Model] = None,
+    signature: Optional[ModelSignature] = None,
+    input_example: Optional[ModelInputExample] = None,
+    pip_requirements: Optional[Union[List[str], str]] = None,
+    extra_pip_requirements: Optional[Union[List[str], str]] = None,
+    metadata: Dict[str, Any] = None,
+    **kwargs,
+) -> None:
+    """
+    Save a trained transformers model to a path on the local file system.
+
+    :param transformers_model: A trained transformers `Pipeline`, `PreTrainedModel`,
+                               `TFPreTrainedModel`, or `FlaxPreTrainedModel` that can be loaded
+                               using the `from_pretrained` method of the respective class.
```

Comment:
I thought we were only accepting pipeline instances as inputs. What changed?

Reply:
There were a few things that popped up that made the Pipeline-only approach not really a viable option:
1. Not all model types support Pipelines (the big name ones do, though) 
2. Custom models are not guaranteed to support Pipelines unless the user builds a custom Pipeline, registers it to the HF repository, and creates a model card on the repository (this isn't very common for users of models; more for the creators of new model architectures. If a user just wants to strip a few layers off of a pretrained model as customization, this new model instance will break if used with the pipeline API)
3. Pipelines aren't serialized as objects. Their configuration is stored (contract between calling objects within an instantiated pipeline are preserved, the model is saved, and the tokenizer / feature extractor / image processor configuration is saved. Supporting passing in a model with a task type is equivalent to supporting a pipeline input with task type from an 'as saved' perspective. 

Mostly because of item 3 above, in order to validate and reconstruct a Pipeline object from the as-saved state we have to do all of the type validation and loading of required tokenizers and processing object elements anyway, the only real change to support a model being passed in (which is functionally locked to a particular tokenizer type / processing object type anyway) was the addition of a few instance validation checks on the saving side. 

We can remove the support for models on input, but it eliminates the utility of this for a non-insignificant number of use cases of the library. 

As to the processor details:
The naming is confusing in the library. 

ImageProcessor -> for vision tasks or multi-modal tasks (purely encodes image data or fetches an image from a url)
FeatureExtractor -> catch-all (sort of) where some vision models can define this in place of an image processor (mostly older ones), is also used to convert audio files to tensors, and can be configured for some LLMs to extract sequences from large bodies of text. multi-modal types also generally (but not always) have these. 
Tokenizer -> exclusively for text -> tensor or tensor -> text

Processor -> a mix of one, some, or all of the above. More advanced logic is supported with these, but they're not compatible with PIpelines. Many use cases of Vision, Audio, MultiModal (and some LLMs) require the use of these. 

The other design option that supports individually saving each of the components from the above set (ImageProcessor, Tokenizer, FeatureExtractor) would be a dict approach for saving via the "model" argument. Since each of those can be successfully put within a Pipeline, it stands to reason to include them in that arg. I'm not so sure if we need to support that 'advanced' functionality right now, though. Because of how deeply coupled these objects are to their respective model architectures, the utility of creating a 'custom' one to interface with a pretrained model is suspect at best. If someone is going to create a new architecture, they'll register the model with all required processing objects to the HF repo.

Reply:
The only caveat to the last point is regarding custom vocabulary files for Tokenizers. I think we should definitely support saving these off, but perhaps that's a future feature to fully support saving customizations (perhaps when we support saving a Trainer object as well)?

Reply:
> We can remove the support for models on input, but it eliminates the utility of this for a non-insignificant number of use cases of the library.

Which use cases? Does it have any impact on LLM use cases? If not, I'm inclined to start with pipeline for now and expand later.

Reply:
Based on the testing this AM, the model-only approach is super dangerous. I've punted that over to be explicit dict-based inputs to support correctness guarantees on retrained models / modified tokenizers. All of the inference logic that was causing complexity and potentially broken predictions is removed.